### ISSUE TITLE
FT-3 — All-species weight saving, new-species initialization spec, and stage-callable fine-tuning

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1582]
**Depends on:** FT-1 (#1580), FM-00 (#1565) · **Blocks:** FT-4 (#1637), MAG-1 (#1584)

**Prerequisite:** the Phase 0 characterization suite is committed — goldens, fixtures and the single-source tolerance table under `tests/golden/`, characterization tests under `tests/unit/`. References of the form `P0-N` name that work.

**Context:** Legacy fine-tuning **shrinks** the model's element set by default: the element table is the union of the training heads' elements (`mace/cli/run_train.py:492-495`), and unless `--foundation_model_elements` is set (default False, `mace/tools/arg_parser.py:633-637`) the foundation model's per-element weights are sliced down to that subset in `load_foundations_elements` (`mace/tools/finetuning_utils.py:27`, dispatching to `load_foundations_elements_default` at `:57` or `load_foundations_elements_magnetic` at `:470` — per-element rows selected by `indices_weights`, every sliced tensor rescaled by `(num_species_foundations / num_species) ** 0.5` at `:90,:150,:160,:183,:202`). The opt-in keep-all path swaps the z-table for the FM's (`run_train.py:496-497`) and **zero-pads** missing E0 entries (`run_train.py:578-586`, `head_energies.get(z, 0.0)`). A third flag, `--foundation_filter_elements` (default True, `arg_parser.py:1046-1055`), is a deprecated alias of `--foundation_model_readout` and shares its dest: it never filtered elements, it decides whether the foundation readout weights transfer, and the call reads the shared dest as `load_readout=args.foundation_model_readout` (`mace/tools/model_script_utils.py:222`). v1 carries the meaning, not the old spelling. A fourth transfer path now exists for the dielectric family — `load_foundations_mdp` (`finetuning_utils.py:646`), which applies the *same* `species_scale` slicing but also transfers higher-order `skip_tp` irreps and all angular-momentum channels in the products — so "the subset path" is three implementations, not one. v1 inverts the default: **fine-tuned artifacts keep the foundation model's full element table by default**, so a model fine-tuned on Si/O today can be fine-tuned again on Fe tomorrow without re-converting; and it goes further — the element table can be **extended beyond** the FM's species, which requires a specified (currently unspecified) initialization for the new per-element parameters. FM-00 (#1565) is a dependency because extension operates on the neutral-format artifact (canonical tensors in safetensors + a JSON config sidecar, never a backend `state_dict`): per-element axes exist there in one pinned canonical layout (e.g. the flat `[Z, A, mul]` symmetric-contraction tensor), so "add a species" is a well-defined row-wise operation. FT-1 (#1580) is a dependency: the third scope item — power users invoking fine-tuning stages independently — builds on FT-1 (#1580)'s heads-as-modules pipeline. Legacy is read in place as the behavioural oracle for the subset path.

**Interface & constraints:**

1. **All-species saving is the default.** Fine-tuning from a (converted) foundation model preserves the FM's full element table in the saved artifact; training on a subset of elements never shrinks it. Both legacy behaviours stay **expressible** as explicit config (the subset-slicing of `--foundation_filter_elements` and the keep-all of `--foundation_model_elements` are KEEP dispositions), but subsetting is a deliberate choice, not the default, and the legacy zero-pad of E0s is replaced by the explicit policy in point 4.

2. **Element-axis extension is a pure artifact operation.** Extending the element table (`extend_elements(artifact, new_zs, init_spec)`-shaped, exact signature at implementation time) acts on the canonical tensors of the neutral format: every per-element parameter (node-embedding rows, the flat `[Z, A, mul]` symmetric-contraction weights, the E0 table, and any other species-indexed tensor the canonical layout declares) gains rows for the new species; **existing rows are copied bit-exact — never rescaled or renormalized**. (The legacy `(num_species_foundations / num_species) ** 0.5` rescale belongs to the subset-slicing path only, where it is applied as a *division* to five separate tensor families; on the keep-all path the factor is 1.)

3. **New-species initialization is a written spec, not an ad-hoc choice.** This ticket authors and implements it. Binding requirements: (a) deterministic — spec + seed reproduce the new rows exactly; (b) **non-perturbing** — a model extended with new species produces bit-identical fp64 outputs on structures containing only the original species (per-element parameters are gathered by species, so unused rows must not leak into any output); (c) recorded — the artifact metadata records the init spec name, seed, and the list of added elements.

4. **E0s for elements outside the FM are a policy, never a crash or a zero default** (shared contract with FT-2 (#1581)): `missing = "error"` (default) raises; `missing = "average"` solves the least squares **restricted to the unresolved elements with the FM values held fixed** (what composite users approximate by hand today); `missing = "zero"` is explicit and validated at data-load time — a config containing a zero-defaulted element under a foundation-sourced E0 spec is a hard error (legacy trains the pt_head against the `head_energies.get(z, 0.0)` pad or crashes late). E0 values come from the artifact's JSON metadata, never from an `atomic_energies_fn` buffer by reflection; any re-estimation writes a **new** `e0s` metadata block with provenance — the FM's own block is immutable.

5. **Stage-callable fine-tuning.** Fine-tuning is a multistage process; a power user can invoke each stage independently, from Python and from the CLI, with typed inputs/outputs and a persistable artifact at every boundary: replay-configuration selection/subselection (FT-1 (#1580)'s assembly + FPS port), pseudolabel generation (FT-2 (#1581)'s `generate_pseudolabels(model, configs, spec)` — pure w.r.t. inputs, `torch.no_grad()`, any per-batch failure raises), element-table extension (this ticket), head/data assembly, and training. **Composition invariant:** running the stages one by one from persisted intermediates produces the same final artifact as the one-shot `mace train` fine-tune path (fp64 tolerance, fixed seeds). FT-3 (#1582) owns the callable surface and its contracts; it does not reimplement the FT-1 (#1580)/FT-2 (#1581) stage internals.

**Task:**
1. Flip the default in the finetune config section: keep the FM's full element table; expose subset-slicing and keep-all as explicit options replacing the `--foundation_filter_elements` / `--foundation_model_elements` pair (one coherent field, not two interacting booleans; the `load_readout=` cross-wiring at `model_script_utils.py:219` has no v1 counterpart).
2. Implement element-axis extension on the neutral-format artifact per constraint 2, covering every species-indexed canonical tensor.
3. Write the new-species initialization spec (docs page + code), satisfying constraint 3; record spec/seed/added-elements in artifact metadata.
4. Wire the `missing` E0 policy for added/out-of-FM elements per constraint 4, including the data-load-time validation for `missing="zero"` and the immutable-FM-metadata rule.
5. Expose the fine-tuning stages as independently callable units (Python API + CLI subcommands) with typed, persistable boundary artifacts; add the composition-equivalence test.
6. Pin the legacy subset path with a parity test against `load_foundations_elements` slicing (including the `species_scale` rescale) on a P0-1 anchor, so the explicit subset option reproduces legacy numbers. The magnetic (`load_foundations_elements_magnetic`) and dielectric (`load_foundations_mdp`) variants of the same slicing are pinned by MAG-1 (#1584) and FT-4 (#1637) respectively; this ticket owns the default path and the shared rule.

**Out of scope:** pseudolabel generation itself and fine-tune shortcuts/presets (FT-2 (#1581)); replay assembly, FPS subselection, LoRA/freeze (FT-1 (#1580)); the MDP transfer path (FT-4 (#1637)) and the magnetic one (MAG-1 (#1584)); the foundation-model converters (FM-1 (#1586)/FM-2 (#1587)); a general model-registry query/extension API beyond adding species (later Axis-O work).

**Acceptance criteria:**
- [ ] Fine-tuning from a converted FM anchor on a two-element subset saves an artifact whose element table equals the FM's full table (default path, no flags).
- [ ] Extending an artifact with new species leaves fp64 outputs on original-species-only structures bit-identical; existing canonical tensor rows are byte-equal before/after extension.
- [ ] Same init spec + seed → identical new-species rows; metadata records spec name, seed, and added elements; the FM's own metadata block is unchanged.
- [ ] `missing="error"` raises on an out-of-FM element; `missing="average"` reproduces the restricted least squares with FM values fixed; `missing="zero"` hard-errors at data load when a zero-defaulted element appears under a foundation-sourced E0 spec.
- [ ] Each fine-tuning stage runs independently from persisted inputs, and the staged composition matches the one-shot fine-tune run at fp64 tolerance with fixed seeds.
- [ ] The explicit subset option matches legacy `load_foundations_elements_default` output (incl. the `(num_species_foundations / num_species) ** 0.5` rescale on all five tensor families) on a P0-1 anchor.

**Inventory gaps assigned here:**
- [ ] `--foundation_filter_elements` / `--foundation_model_elements` — all-species saving is a v1 default, so the filtering behaviour needs pinning before it changes. `tests/unit/test_finetuning_utils.py` exists but does not cover the flag pair end to end.

**Verify:**
```bash
python -m pytest packages/mace-torch/tests/finetune/test_all_species.py -v        # default table, extension, init spec
python -m pytest packages/mace-torch/tests/finetune/test_stage_callable.py -v     # per-stage invocation + composition
python -m pytest tests/parity -m "not slow" -k "species or finetune"              # subset path vs frozen legacy
python -m pytest tests/unit/test_finetuning_utils.py -v
```
**Review focus:** the non-perturbation invariant (extension must be provably inert for existing species — check every species-indexed tensor, not just the embedding), the three `missing` E0 policy branches and their validation timing, and that no stage boundary leaks an untyped namespace.

---
